Introduce concurrent tasks scheduler package

talhadar1 opened the pull request:
## **🚀 First Pull Request at OneLayer!**

This pull request is my first at OneLayer! 🎉

## **Overview**

This PR introduces a comprehensive **scheduler package** that enables concurrent task execution with advanced control features. 
The package provides a robust foundation for running tasks concurrently while maintaining visibility into their execution status.

## **Key Features**

### **Core Scheduler**

- **Concurrent Execution**: Configurable worker pool with controllable concurrency limits
- **Task Registration**: Simple interface for registering tasks with optional decorators
- **Real-time Monitoring**: Live tracking of ongoing, finished, and failed tasks
- **Comprehensive Reporting**: Detailed summary reports with success/failure statistics

### **Task Decorators (Higher-Order Functions)**

The package implements a flexible decorator pattern allowing tasks to be enhanced with additional capabilities:

- **`WithName`**: Assigns custom names to tasks for better identification
- **`WithLogging`**: Adds detailed execution logging
- **`WithRetry`**: Implements automatic retry logic with configurable retry counts
- **`WithDelay`**: Adds scheduled delays before task execution

### **Task Interface**

Clean, simple interface that any task can implement:

```go
type RegisterableTask interface {
    Run(ctx context.Context) error
}
```

## **Architecture Highlights**

- **Worker Pool Pattern**: Efficient concurrent execution with configurable max workers
- **Graceful Shutdown**: Proper cleanup and completion tracking

## **Usage Example**

```go
scheduler := NewScheduler()
scheduler.SetMaxWorkers(10)

// Register tasks with decorators
scheduler.RegisterTask(yourTask,  
    1,  // task-id
    WithName("your-task-name"),            
    WithRetry(3),
    WithLogging(),
    WithDelay(5*time.Second),
)


go scheduler.RunScheduler() // execute all tasks

scheduler.Stop()   // stop accepting new tasks and wait for all registered tasks to finish 

scheduler.SummaryReport()
```

Proof: https://github.com/talhadar1/scheduler/pull/1#issuecomment-3004647682

talhadar1 committed:
task-scheduler: introduce core scheduling API and task interface

This commit introduces the foundational structure for a concurrent task
scheduler. It defines the RegisterableTask interface for units of
executable work and a task struct that wraps this interface with
additional metadata such as ID and name. The Scheduler struct manages
task execution using buffered channels for queuing (taskChan) and
signaling completion (doneChan). The commit also includes method stubs
for RegisterTask, RunScheduler, and Stop, establishing the control flow
for task submission and lifecycle management. These components lay the
groundwork for a robust, extensible scheduling system that supports
concurrent execution patterns.

talhadar1 commented:
**Proof Of Work:**
**short example of main.go**
![image](https://github.com/user-attachments/assets/f1be2d04-5ba6-4c55-b5ac-5d2ad44027d3)

**output:**
![image](https://github.com/user-attachments/assets/6ed3e089-49b8-4cb2-a9de-e208da487352)


📌 Key Observations from Logs

1. Delay is working as you can observe task-1 registered before task-2 but  task-2 runs before task-1 that is because task-1 was delayed(100ms)  more than task-2(50ms)
2. Task 3 demonstrat retry 
3. Tasks 3-4 demonstrate naming - (`retryTask` & `TAL-Error-Task`)
4. Logging - task-0 (no logging decorator) has no output but it can be viewed that this task was part of the scheduler run as it counts 5 registered tasks (4 with logging enabled and 1 without)
5. Real time monitoring - in the out put you can observe the blue lines that represent the ongoing scheduler status
![image](https://github.com/user-attachments/assets/47d3d734-7d6e-4645-9581-b834f65b5b8f)
![image](https://github.com/user-attachments/assets/960a8573-b0e6-41ad-b36f-92813771fe2c)

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -0,0 +1,73 @@
+// Task is the unexported implementation of the RegisterableTask interface,
+// allowing additional metadata and behavior (retry count, delay, logger)
+// to be wrapped around an inner RegisterableTask.
+type Task struct {
```
The comment says it's unexported, but it is.

talhadar1 replied:
changed to un-exported as the comment says

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -0,0 +1,73 @@
+		NumOfRetries: 1,   // default no retries
+	}
+	return newTask
+}
```
Should this be exported?

GalShnapp replied:
Also, I'd probably add this constructor in the commit I use it - the task registration commit.

talhadar1 replied:
no need to export this is for internal use.

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -0,0 +1,73 @@
+type Scheduler struct {
+	taskChan   chan Task
+	doneChan   chan struct{}
+	maxWorkers int
```
Provide some docs about these fields?

talhadar1 replied:
Done

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -48,13 +49,23 @@ func (t *Task) Run(ctx context.Context) error {
+	registeredTasks int
+	ongoingTasks    int
+	failedTasks     int
+	finishedTasks   int
```
These have more to do with monitoring than scheduling tasks. Do they really belong in this commit?

talhadar1 replied:
Correct, we use those mainly for monitoring proposes, but I also use them in the unit-tests of the RunScheduler(), in my opinion to insert them at the last commits with the monitoring is adding some "logic" to the main runner function and new unit-tests for them is not convenient. 
What do you think?

talhadar1 replied:
`registeredTasks` , `ongoingTasks` , `failedTasks` , `finishedTasks` are introduced with the scheduler logic as my comment above explain.
other fields introduced in the commits that contains the relevant logic for them.

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -68,6 +79,50 @@ func NewScheduler() *Scheduler {
+	close(s.taskChan)
+}
+
+// Done returns a channel that's closed when all tasks have finished.
```
Add a small explanation on how and why should a developer use this, like you did with `func (s *Scheduler) Stop()`.

Something like `Call this function and wait on it's channel to assert the task queue is empty.

talhadar1 replied:
For simplicity I will merge Done() into Stop()
The new Stop() will close the scheduler from registering new tasks and wait for all tasks to finish (also delayed tasks)

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -36,6 +36,10 @@ func NewTask(inner RegisterableTask, id int, opts ...TaskOption) *Task {
+	// Apply any TaskOption to the new Task
+	for _, option := range opts {
+		option(newTask)
+	}
```
This commit is a bit small. I'd probably add it inside the commit the present's the registration logic.

talhadar1 replied:
squashed this change with the commit that present's the registration logic.

GalShnapp commented on task-scheduler/scheduler_test.go:
```diff
@@ -110,11 +110,12 @@ func TestScheduler_RegisterTask(t *testing.T) {
 			name:                    "register single task",
 			tasks:                   []RegisterableTask{mocks.NewMockTask(1, "test-task")},
-			options:                 [][]TaskOption{{}},
+			options:                 [][]TaskOption{},
```
What happened here?

talhadar1 replied:
fixed

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -98,7 +98,9 @@ func (s *Scheduler) RunScheduler() {
-
+				if task.delay > 0 {
+					time.Sleep(task.delay) // Delay before running the task
+				}
```
This solves the delay issue - but now one of your workers is asleep ;)

talhadar1 replied:
added delay logic:
new delay dispatcher is add, when registering a delayed task, it will move to the dispatcher to handle.
when the time arrive to run the task the dispatcher will add the task to the tasks channel to be executed.
This added one more go routine for the dispatcher but keeps all workers 'alive' and working on ongoing tasks.

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -0,0 +1,73 @@
+	inner        RegisterableTask // wrapped RegisterableTask that provides the actual Run logic
+	logger       *log.Logger      // optional logger for Task-specific output
+	delay        time.Duration    // delay duration before execution
+	NumOfRetries int              // number of times to retry on failure
```
I'd add each of these fields with the commit that enables a user to use them.

talhadar1 replied:
done

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -166,3 +167,18 @@ func (s *Scheduler) Done() <-chan struct{} {
+		)
+		time.Sleep(d)
+	}
+}
```
This is just epic. I love the use of "\033[0m".

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -161,3 +161,8 @@ func (s *Scheduler) Stop() {
+// SetMaxWorkers configures the maximum number of concurrent tasks.
+func (s *Scheduler) SetMaxWorkers(maxWorkers int) {
+	s.maxWorkers = maxWorkers
+}
```
If this is the commit where you enable a use to alter the number of workers up to this point, I wouldn't have used it as a property of the scheduler struct beforehand. I'd use a constant before and add it as a property here.

talhadar1 replied:
Done

GalShnapp reviewed (changes requested):
Great work. 

You did a really good with the user interface - providing a clear API to the user and a great experience with your monitoring abilities.

You also did great work diving your changes into small, readable change sets that are reviewable - I know it's not easy, so well done.


I've got one general note, and a few other more specific notes the will also be added.
You've written your commit messages in a change log format, using bullets to convey the changes you've made. This doesn't really go along with our contribution style and development guideline. Instead, write in a summary of what you did and why. If you'd like, we can schedule a meeting where we'd write some commit messages together.

talhadar1 committed:
task-scheduler: add worker pool and atomic counters for task execution

This commit completes the initial implementation of the concurrent task
scheduler. It introduces a worker pool (RunScheduler) that runs up to 50
 tasks concurrently using goroutines. Tasks are submitted through
 RegisterTask, wrapped with metadata using newTask, and tracked using
 atomic counters for finished, failed, and ongoing executions. added
 Stop method ensures a graceful shutdown once all tasks are processed.

A MockTask type is added under the mocks package to simulate real-world
task behavior during tests.

talhadar1 committed:
task-scheduler: log tasks using decorator

This commit introduces a WithLogging decorator that enables structured
logging for individual tasks. When applied, it injects a task-specific
logger that outputs to stdout, prefixing each log entry with the task ID
 and name. The task struct is extended with a logger field to support
 this behavior.

Logging is integrated into both the task registration and execution
phases. When a decorated task is registered, it logs its registration
metadata. During execution, it logs the start, success, or failure of
the task, providing enhanced observability for debugging and analysis.

This enhancement improves the traceability of tasks during concurrent
execution and sets the foundation for future decorators like retry or
timeout handling.

talhadar1 committed:
task-scheduler: name tasks using decorator

This commit introduces the WithName decorator, allowing callers to
override a task’s default auto-generated name with a custom one. This
improves task identification in logs and test assertions, especially
when combined with the existing WithLogging decorator.

If WithLogging is also applied, the logger prefix is updated to reflect
the overridden task name. The decorator is idempotent and respects the
order of application—later decorators take precedence in naming.

talhadar1 committed:
task-scheduler: delay tasks using decorator

This commit introduces the WithDelay decorator, allowing tasks to be
scheduled with a specified execution delay. A new component,
delayDispatcher, manages all delayed tasks using a single goroutine and
ensures they are enqueued into the scheduler at the correct execution
time. This design avoids spawning separate goroutines per delayed task
and ensures precise, efficient time-based scheduling.

The task struct is extended with a delay field, and RegisterTask now
routes delayed tasks through the dispatcher while immediate tasks go
directly to the execution channel. The scheduler’s Stop method now waits
 for all delayed tasks to be flushed before closing the main task
 channel.

talhadar1 committed:
task-scheduler: retry tasks execution on failure using decorator

This commit introduces retry functionality to the scheduler through the
new WithRetry decorator. Tasks can now be configured to automatically
retry upon failure up to a defined limit (between 1 and 100). The task
struct includes a NumOfRetries field, and RunScheduler is updated to
loop task execution attempts until success or exhaustion of retries.

To support robust testing, a RetryTestTask has been added to the mocks
package. It simulates failures and tracks retry attempts, allowing
precise validation of the retry logic.

task-scheduler: retry tasks execution on failure using decorator

This commit introduces retry functionality to the scheduler through the
new WithRetry decorator. Tasks can now be configured to automatically
retry upon failure up to a defined limit (between 1 and 100). The task
struct includes a NumOfRetries field, and RunScheduler is updated to
loop task execution attempts until success or exhaustion of retries.

To support robust testing, a RetryTestTask has been added to the mocks
package. It simulates failures and tracks retry attempts, allowing
precise validation of the retry logic.

talhadar1 committed:
task-scheduler: enable amount of tasks workers modification

This commit introduces the ability to configure the number of concurrent
 worker goroutines used by the scheduler. A new SetMaxWorkers method
 allows users to specify the desired level of parallelism before task
 execution begins. If not explicitly set, the scheduler defaults to
 defaultNumOfWorkers.

Internally, RunScheduler checks the maxWorkers field and falls back to
the default if an invalid value (≤ 0) is provided. This flexibility
allows developers to fine-tune the concurrency level to match workload
characteristics or resource constraints.

talhadar1 committed:
task-scheduler: introduce SchedulerMonitor for live runtime insights

This commit adds a SchedulerMonitor method to periodically report the
internal state of the scheduler. It prints key metrics including
ongoing, registered, delayed, finished, and failed tasks, as well as the
 number of active goroutines. The output is color-coded for improved
 visibility in terminal environments.

This enhancement improves runtime observability, helping developers and
operators monitor task progress, system load, and concurrency behavior
in real time.

talhadar1 committed:
task-scheduler: summerize scheduler work

This commit adds the SummaryReport() method to the scheduler, enabling a
 concise summary of task execution results. It prints the total number
 of registered, successful, and failed tasks, and should be called after
 Stop() to ensure all processing has completed.

talhadar1 committed:
go.mod: get github.com/stretchr/testify

talhadar1 committed:
main.go: add main entry point and example tasks

This commit introduces a runnable main.go file along with an example
package containing sample task implementations. These include a basic
task (Task) and a simulated error-producing task (ErrorTask). The main
program demonstrates how to use the scheduler with various combinations
of decorators such as WithLogging, WithDelay, WithRetry, and WithName.

It also includes a real-world usage pattern by registering a variety of
tasks—some delayed, some retry-enabled, and others with logging, then
launching the scheduler and printing a final summary report. This setup
serves both as a usage example and a practical integration test
environment.

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -0,0 +1,58 @@
+
+// task is the unexported implementation of the RegisterableTask interface,
+// allowing additional metadata id and name
+// to be wrapped around an inner RegisterableTask.
```
Comments should start with a capital letter, since they're in english.

Rephrase this as `Struct task` or `A task` with structs. When using functions, use `Call foo()`

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -0,0 +1,58 @@
+// and sets the maximum number of concurrent workers.
+func NewScheduler() *Scheduler {
+	return &Scheduler{
+		taskChan: make(chan task, 100),
```
Use a const for this.

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -29,12 +50,21 @@ func (t *task) Run(ctx context.Context) error {
 type TaskOption func(*task)
 
 // RegisterTask submits a Task for execution.
```
Note that since you used channel for this, this function is blocking. I'd document this As this is important to note.
Also, s.registeredTasks is not thread safe.

talhadar1 replied:
- Changed s.registeredTasks to a thread safe type. 
- Added the relevant blocking comment

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -48,11 +78,40 @@ func NewScheduler() *Scheduler {
-func (s *Scheduler) RunScheduler() {}
+// It drains Scheduler taskChan until closed, then waits for all workers.
+// Workers run tasks concurrently, updating the counts of ongoing, finished, and failed tasks.
+// It uses a mutex to protect access to the Scheduler counters.
```
It doesn't use mutexes anymore

GalShnapp replied:
It's blocking and should be used in a separate routine, and you should document it.

GalShnapp commented on task-scheduler/delayDispatcher.go:
```diff
@@ -0,0 +1,108 @@
+	}
+}
+
+// maxDuration guards against negative or huge waits.
```
just negatives

GalShnapp commented on task-scheduler/delayDispatcher.go:
```diff
@@ -0,0 +1,108 @@
+			timer.Stop()
+			timer = nil
+			timerCh = nil
+		}
```
Some docs would be great here.

GalShnapp commented on task-scheduler/delayDispatcher.go:
```diff
@@ -0,0 +1,108 @@
+			}
+			queue = append(queue, delayedTask{}) // make space
+			copy(queue[i+1:], queue[i:])         // shift elements to the right
+			queue[i] = req                       // insert the new task
```
You should probably implement a Q elsewhere.

GalShnapp commented on task-scheduler/scheduler.go:
```diff
@@ -51,35 +54,48 @@ func (t *task) Run(ctx context.Context) error {
 // It takes an existing RegisterableTask and returns a new, decorated RegisterableTask.
 type TaskOption func(*task)
 
-// RegisterTask submits a Task for execution.
```
Please fix this in the commit this comment was added.

GalShnapp commented on task-scheduler/decorators.go:
Couldn't the whole scheduler act this way, with the time for undelayed tasks being before this queue was initiated? Does this really have to be a different dispatcher?

talhadar1 replied:
I decided to take the delayed task scheduling and separate it to a different logic block, as in my opinion it not directly relevant for the worker execution and the task time sorting should not be a part of the RunScheduler()  logic. 

I agree that we could have enabled the dispatcher logic on all of the registered tasks and send all of the tasks to a priority queue that the worker will pull from it.

The delay task is a requirement that I have added to the scheduler and was not required as part of the original task.
As we discussed I agree that much more work can be done to optimize this solution.

GalShnapp reviewed (changes requested):
Keep up with the good work.
Two points:
- We want docs to start with capital letters and obey the english language's grammar. I wrote this once, but expect this to be changed all over the code. 
- Is the delayed dispatcher really needed?

talhadar1 commented:
Thanks Gal! 
I have fixed all the issues in your comments.
For the issue you raised, splitting the logic of the delayed dispatcher I have answered in [this comment](https://github.com/talhadar1/scheduler/pull/1#discussion_r2173737511). 

Thank you for your time and guidance 💚
